File: samples/java-valueentity-store/src/main/java/store/product/domain/ProductValueEntity.java
package store.product.domain;

import com.google.protobuf.Empty;
import kalix.javasdk.valueentity.ValueEntityContext;
import store.product.api.ProductApi;

public class ProductValueEntity extends AbstractProductValueEntity {

  public ProductValueEntity(ValueEntityContext context) {}

  @Override
  public ProductDomain.ProductState emptyState() {
    return ProductDomain.ProductState.getDefaultInstance();
  }

  @Override
  public Effect<Empty> create(ProductDomain.ProductState currentState, ProductApi.Product product) {
    ProductDomain.ProductState productState =
        ProductDomain.ProductState.newBuilder()
            .setProductId(product.getProductId())
            .setProductName(product.getProductName())
            .setPrice(
                product.hasPrice()
                    ? ProductDomain.Money.newBuilder()
                        .setCurrency(product.getPrice().getCurrency())
                        .setUnits(product.getPrice().getUnits())
                        .setCents(product.getPrice().getCents())
                        .build()
                    : ProductDomain.Money.getDefaultInstance())
            .build();
    return effects().updateState(productState).thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<ProductApi.Product> get(
      ProductDomain.ProductState currentState, ProductApi.GetProduct getProduct) {
    ProductApi.Product product =
        ProductApi.Product.newBuilder()
            .setProductId(currentState.getProductId())
            .setProductName(currentState.getProductName())
            .setPrice(
                currentState.hasPrice()
                    ? ProductApi.Money.newBuilder()
                        .setCurrency(currentState.getPrice().getCurrency())
                        .setUnits(currentState.getPrice().getUnits())
                        .setCents(currentState.getPrice().getCents())
                        .build()
                    : ProductApi.Money.getDefaultInstance())
            .build();
    return effects().reply(product);
  }
}

File: samples/scala-valueentity-store/src/main/scala/store/product/domain/ProductValueEntity.scala
package store.product.domain

import com.google.protobuf.empty.Empty
import kalix.scalasdk.valueentity.ValueEntity
import kalix.scalasdk.valueentity.ValueEntityContext
import store.product.api

class ProductValueEntity(context: ValueEntityContext) extends AbstractProductValueEntity {

  override def emptyState: ProductState = ProductState.defaultInstance

  override def create(currentState: ProductState, product: api.Product): ValueEntity.Effect[Empty] = {
    val productState = ProductState(
      productId = product.productId,
      productName = product.productName,
      price = product.price.map(money => Money(money.currency, money.units, money.cents)))
    effects.updateState(productState).thenReply(Empty.defaultInstance)
  }

  override def get(currentState: ProductState, getProduct: api.GetProduct): ValueEntity.Effect[api.Product] = {
    val product = api.Product(
      productId = currentState.productId,
      productName = currentState.productName,
      price = currentState.price.map(money => api.Money(money.currency, money.units, money.cents)))
    effects.reply(product)
  }
}
